### PR TITLE
Retry indefinitely for s3 indices blob read errors

### PR DESCRIPTION
**Pull request by ywangd**
This PR makes the s3 readBlob to retry indefinitely on either opening or reading errors when the operation purpose is Inidces. It perform retries with no delay within the configured number of max retries. Beyond that, it retries with increased delay each time with a capped maximum of 10 seconds.

Relates: ES-6453

### CODE REVIEW

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:276]: The choice of initial and max delay is modelled after the ones in [IndicesService](https://github.com/elastic/elasticsearch/blob/d47fee26644ad93a75c0a83a62c3eee14d349d0f/server/src/main/java/org/elasticsearch/indices/IndicesService.java#L1375-L1376)

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (kingherc) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:242]: The "maximum number of read attempts" was previously `blobStore.getMaxRetries() + 1` (a constant) and now it seems to me to be variable?

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (kingherc) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:276]: nit I keep thinking about the fact that you mentioned we sleep in a blocking mode typically. I wonder if maybe 2 seconds maximum is more suitable than 10 seconds. But I am also not sure why; just a hunch.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (kingherc) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:196]: nit All this code up to this point also appears similarly above below line 105. I wonder if maybe this should be refactored into another function like `checkCanRetry(action, e)` or something like that.

File: modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (kingherc) [inline modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java:670]: I guess this controls the difference between an attempt with meaningful progress and one without meaningful progress?

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: I extracted a new method `recordAndLogForRetry` for these and the increment of `attempt`.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:275]: Changing it to 2 seconds does not really change the blocking nature. The thread just wakes up more often, potentially a bit earlier to pick up if s3/network recovers. As commented [above](https://github.com/elastic/elasticsearch/pull/103300#discussion_r1423368679), I chose 10 seconds after what IndicesService does. If we do get to 10 seconds, the situation is probably _not_ going to resolve very quickly. So waiting a bit longer seems OK.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: Thanks for catching this. I mis-read the message. Pushed cb640128a22ea738f68164a33c927d59da02a2d5

File: modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (ywangd) [reply modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java:670]: No this controls whether it makes any progress, meaningful or not, at all. The server just drops the connection.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (kingherc) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:180]: nit You could potentially put this `if` inside `recordAndLogForRetry`, but I also do see the value of it being blatantly here for code readability.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (kingherc) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:274]: nit

```suggestion
    // protected access for testing
```

**Review by kingherc — APPROVED** [review]
LGTM. But it'd be good to get a second approval before merging.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: I moved this block into `recordAndLogForRetry` as well and renamed the method to `MaybeRecordAndLogForRetry`. Thanks!

**Comment by ywangd** [comment]
@elasticmachine update branch

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (volodk85) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:189]: Would it be possible emit apm metric always when we retry? 
I'd like to have a sense how many retries do we do (avg, count)

Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:189]: It could a useful metric specific to s3. I can work on it in a separate follow-up. Thanks for the idea.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (henningandersen) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:191]: Can we ensure the first retry is always output on INFO? Perhaps something like:

`Integer.bitCount(attempt) == 1`?

which would log exponentially less.

**Review by henningandersen — COMMENTED** [review]
Looks good, left a few comments. Did not read the tests, assuming Iraklis reviewed that part.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:276]: This is to keep the existing behaviour which retries read immediately within the configured `maxAttempts`.  Therefore the new behaviour adds delay only when the configured `maxAttempts` is exhausted. I expanded the comment with this explanation. I hope it helps.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:263]: I was thinking that the interuption could go suppressed and just proceed with another round of retry. On a second thought, your suggestion makes sense and I adopted it.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:245]: Yeah I think it can be a follow-up. Existing issues that I have studied so far are all for indices data. Failure to read translog data may have a different failure pattern. I can look into them after this.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:191]: I updated it to log the very first retry with
```
logForRetry((attempt == 2 || attempt % 30 == 0) ? Level.INFO : Level.DEBUG, action, e);
```
Note the `attempt` starts with `1` as it is a count for all attempts including the original one and subsequent retries so that the first retry is the 2nd attempt.

I prefer to keep logging about every 5 minutes since the behaviour would be more predicatble in cases where we experience extended period of s3/network outages.

**Comment by ywangd** [comment]
> It could be nice to also log when the S3 "problem" goes away, but we can look at that as future work.

We could check whether `attempt` has increased when the read methods or the constructor returns and log a message if it does. This would tell us that a particular read has overcomed the S3 problem. I'll get to it as a follow-up. Thanks for the suggestion.

**Comment by ywangd** [comment]
@elasticmachine update branch

**Comment by ywangd** [comment]
Gentle reminder for review. We had another page for s3 being unreachable (`java.net.UnknownHostException`) over the weekend. I'd like to get this merged to see whether we can get rid of these issues. Thanks!

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (henningandersen) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: I'd prefer to harmonize the delays of existing and new retries, but I am also ok to postpone that to a follow-up. It just seems that the code is more complicated than it should be now.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (henningandersen) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: I am ok with that, but notice that if there is an s3 outage, we could have many threads doing this per node, perhaps 50 or so currently, perhaps more in the future for large nodes. Still just a message every 6s (per node), so feels ok'ish. We can refine later too ofc.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (henningandersen) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:198]: Should this be `attempt == 2`? Otherwise I think we'll not get there?

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (henningandersen) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:189]: Would it not be more intuitive to increment this after `logForRetry` and avoid subtracting 1 from attempt there?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: I have come around to your point. The same delay is now applied for all retries.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: Potential excessive logging from multiple threads is a good point. I can adjust this again in the follow-up where I will also add log messages when the retry eventually succeeds (see also https://github.com/elastic/elasticsearch/pull/103300#issuecomment-1882130933).

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: It is `attempt == 1` because this branch is used to log failures where no retry should be attempted, i.e. repository analysis. I see now it can be confusing. So I separate this failure logging into a separate method and get rid of the check for `attempt`.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java
Comment (ywangd) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java]: This is related to the above comment. I had to increment it first because `attempt == 1` was used to signal failure without retries. It was confusing. I have now made them separate and therefore `attempt` is now incremented after the logging.

**Pull request merged**
